src/main/java/com/contrastsecurity/maven/plugin/sdkx/scan/ArtifactScanner.java

```diff
@@ -0,0 +1,84 @@
+package com.contrastsecurity.maven.plugin.sdkx.scan;
+
+import com.contrastsecurity.exceptions.UnauthorizedException;
+import com.contrastsecurity.maven.plugin.sdkx.CodeArtifact;
+import com.contrastsecurity.maven.plugin.sdkx.ContrastScanSDK;
+import com.contrastsecurity.maven.plugin.sdkx.NewCodeArtifactRequest;
+import com.contrastsecurity.maven.plugin.sdkx.Scan;
+import com.contrastsecurity.maven.plugin.sdkx.StartScanRequest;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Facilitates analyzing code artifacts with Contrast Scan. Uploads the code artifact to Contrast
+ * Scan, starts a new scan, and provides means for retrieving the results of that scan when it has
+ * completed.
+ */
+public final class ArtifactScanner {
+
+  private final ScheduledExecutorService scheduler;
+  private final ContrastScanSDK contrast;
+  private final String organizationId;
+  private final String projectId;
+  private final Duration interval;
+
+  /**
+   * @param scheduler for scheduling scan status retrievals and all other retrievals communication
+   *     with the Scan API
+   * @param contrast for communicating with the Contrast Scan API
+   * @param organizationId unique ID of the organization in which the project exists
+   * @param projectId unique ID of the Contrast Scan project
+   * @param interval the polling interval for retrieving the status of a not-yet-finished scan
+   */
+  public ArtifactScanner(
+      final ScheduledExecutorService scheduler,
+      final ContrastScanSDK contrast,
+      final String organizationId,
+      final String projectId,
+      final Duration interval) {
+    this.scheduler = Objects.requireNonNull(scheduler);
+    this.contrast = Objects.requireNonNull(contrast);
+    this.organizationId = Objects.requireNonNull(organizationId);
+    this.projectId = Objects.requireNonNull(projectId);
+    this.interval = Objects.requireNonNull(interval);
+  }
+
+  /**
+   * Uploads the given code artifact to Contrast and starts a new scan.
+   *
+   * @param file code artifact to analyze
+   * @param label distinguishes this scan from others in your project
+   * @return new {@link ScanOperation} for retrieving scan results
+   */
+  public ScanOperation scanArtifact(final Path file, final String label) {
+    final NewCodeArtifactRequest newCodeArtifactRequest =
+        NewCodeArtifactRequest.of(projectId, file);
+    final CodeArtifact artifact;
+    try {
+      artifact = contrast.createCodeArtifact(organizationId, newCodeArtifactRequest);
+    } catch (final IOException e) {
+      throw new UncheckedIOException("Failed to upload code artifact", e);
+    } catch (final UnauthorizedException e) {
+      throw new IllegalStateException("Failed to authenticate to Contrast", e);
```

Comment:
This is part of the reason I put this in a new package. I wanted these relatively higher level classes to be a little friendlier to use.

The Unauthorized exception is particularly frustrating to handle on every call.

When making a single call with sdkx, then sure, but a class like this one that makes multiple calls on your behalf, I don't want to think about it. I just want to know if the scan completes or not. Does that make sense?

Reply:
But by combining all these calls into a single method, doesn’t that make handling `UnauthorizedException` more user-friendly anyway. You’ll only ever handle it once I think. I prefer to expose potential exceptions to the caller so that they can tell what might go wrong.

But for the plugin, it all ends up stopping the plugin execution, so it doesn’t really matter much. I’m okay with leaving as-is, but maybe we’ll revisit when we migrate the code to the SDK.

Reply:
OK, happy to keep things as-is due to my current time crunch, and we have some things to think about when we revisit this on the SDK migration such as:

* Do users need to distinguish between authentication problems and other problems with the API?
* Should we distinguish between problems that could be resolved with a retry and those that could not?
* Should we use checked exceptions? If yes, is it always or sometimes?